src/python/zquantum/core/wip/evolution.py

```diff
@@ -0,0 +1,192 @@
+        circuit = circuits.import_from_pyquil(pyquil.paulis.exponentiate(term))
+
+        new_circuit = circuits.Circuit(
+            [_adjust_gate_angle(operation, time) for operation in circuit.operations]
+        )
+    else:
+        exponent = term * time
+        new_circuit = circuits.import_from_pyquil(pyquil.paulis.exponentiate(exponent))
+
+    return new_circuit
+
+
+def time_evolution_derivatives(
+    hamiltonian: pyquil.paulis.PauliSum,
+    time: float,
+    method: str = "Trotter",
+    trotter_order: int = 1,
+) -> Tuple[List[circuits.Circuit], List[float]]:
+    """Generates derivative circuits for the time evolution operator defined in
+    function time_evolution
+
+    Args:
+        hamiltonian: The Hamiltonian to be evolved under. It should contain numeric
+            coefficients, symbolic expressions aren't supported.
+        time: time duration of the evolution.
+        method: time evolution method. Currently the only option is 'Trotter'.
+        trotter_order: order of Trotter evolution
+
+    Returns:
+        A Circuit simulating time evolution.
+    """
+    if method != "Trotter":
+        raise ValueError(f"The method {method} is currently not supported.")
+
+    single_trotter_derivatives = []
+    factors = [1.0, -1.0]
+    output_factors = []
+
+    for i, term_1 in enumerate(hamiltonian.terms):
+        for factor in factors:
+            output = circuits.Circuit()
+
+            try:
+                r = complex(term_1.coefficient).real / trotter_order
+            except TypeError:
+                raise ValueError(
+                    "Term coefficients need to be numerical. "
+                    f"Offending term: {term_1}"
+                )
+            output_factors.append(r * factor)
+            shift = factor * (np.pi / (4.0 * r))
+
+            for j, term_2 in enumerate(hamiltonian.terms):
+                output += time_evolution_for_term(
+                    term_2,
+                    (time + shift) / trotter_order if i == j else time / trotter_order,
+                )
+
+            single_trotter_derivatives.append(output)
+
+    if trotter_order > 1:
+        output_circuits = []
+        final_factors = []
+
+        repeated_circuit = time_evolution(
+            hamiltonian, time, method="Trotter", trotter_order=1
+        )
+
+        for position in range(trotter_order):
+            for factor, different_circuit in zip(
+                output_factors, single_trotter_derivatives
+            ):
+                output_circuits.append(
+                    generate_circuit_sequence(
+                        repeated_circuit, different_circuit, trotter_order, position
+                    )
+                )
+                final_factors.append(factor)
+        return output_circuits, final_factors
+    else:
+        return single_trotter_derivatives, output_factors
+
+
+def generate_circuit_sequence(repeated_circuit, different_circuit, length, position):
```

Comment:
Shouldn't this be private?

Reply:
Also, lacks typing.

Reply:
I left it as a public function because it was public in the old `evolution.py`. However, since no one i using this wip package, it might indeed be a good idea to make it private. This and the lack of typing have been addressed in 432aa86.